neoscore/western/tuplet.py

```diff
@@ -0,0 +1,140 @@
+
+    It is not common music engraving practice to split tuplets
+    across bar lines or staff splits, so this function is
+    not supported.
+    """
+
+    def __init__(
+        self,
+        start: PointDef,
+        start_parent: PositionedObject,
+        end: PointDef,
+        end_parent: Optional[PositionedObject] = None,
+        ratio_text: str = "3",
+        include_bracket: bool = True,
+        bracket_dir: DirectionY = DirectionY.DOWN,
+        font: Optional[MusicFont] = None,
+    ):
+        """
+        Args:
+            start: The starting point.
+            start_parent: The parent for the starting position. If no font is given,
+                this or one of its ancestors must implement :obj:`.HasMusicFont`.
+            end: The spanner end point position. This must be to the right of the start position
+            end_parent: An object either in a Staff or
+                a staff itself. The root staff of this *must* be the same
+                as the root staff of ``start_parent``. If omitted, the
+                stop point is relative to the start point.
+            ratio_text: The Tuplet number (e.g. 3 for tripets) or Ratio number (e.g. 5:4 for polyrhythms)
+            include_bracket: Bool to draw the spanning bracket over the tuplet
+            bracket_dir: The direction the line's ending hook points.
+                For lines above staves, this should be down, and vice versa for below.
+            font: If provided, this overrides any font found in the ancestor chain.
+        """
+        PositionedObject.__init__(self, start, start_parent)
+        Spanner2D.__init__(self, end, end_parent or self)
+        if font is None:
+            font = HasMusicFont.find_music_font(start_parent)
+        self._music_font = font
+        self.direction = bracket_dir
+
+        # Create line path object
+        self.line_path = Path(
+            ORIGIN,
+            self,
+            Brush.no_brush(),
+            Pen(thickness=font.engraving_defaults["tupletBracketThickness"]),
+        )
+
+        # Calculate the bracket end length
+        self.bracket_height = self.music_font.unit(self.direction.value)
+        # Draw bracket
+        if include_bracket:
+            self._draw_path()
+
+        # Convert ratio text to SMuFL glyphs
+        self.smufl_text = self._number_to_digit_glyph_names(ratio_text)
+
+        # Create line text object; will paint over bracket line
+        spanner_center = self.point_along_spanner(0.5)
+        self.line_text = MusicText(
+            (spanner_center.x, spanner_center.y + self.bracket_height),
+            self,
+            self.smufl_text,
+            font,
+            background_brush=neoscore.background_brush,
+            alignment_x=AlignmentX.CENTER,
+            alignment_y=AlignmentY.CENTER,
+        )
+
+    def _draw_path(self):
+        """Draw the path according to this object's attributes.
+
+        Returns: None
```

Comment:
There's no need to specify return types since we use type signatures now that document this for free.